File: java-exercise-fetcher/src/test/java/org/jesperancinha/twitter/service/SchedulerServiceImplTest.java
package org.jesperancinha.twitter.service;

import org.jesperancinha.twitter.client.TwitterClient;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.Mockito;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.test.context.ActiveProfiles;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.only;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;

@SpringBootTest
@ActiveProfiles("scheduler")
public class SchedulerServiceImplTest {

    @Autowired
    private RunningService runningService;

    @MockBean
    private TwitterClient twitterClient;

    @BeforeEach
    public void setUp() {
        Mockito.reset(twitterClient);
    }

    @Test
    public void testStartProcess_whenRunnerCreated_thenSchedulerServiceInstance() throws InterruptedException {
        assertThat(runningService).isInstanceOf(SchedulerServiceImpl.class);

        verify(twitterClient, never()).startFetchProcess();
        verifyNoMoreInteractions(twitterClient);

    }

    @Test
    public void testStartProcess_whenRunnerStarted_thenSchedulerServiceStart() throws InterruptedException {
        runningService.startProcess();

        verify(twitterClient, only()).startFetchProcess();
        verifyNoMoreInteractions(twitterClient);

    }

    @Test
    public void testStartProcess_whenSchedulerStarted_thenSchedulerServiceStarte() throws InterruptedException {
        ((SchedulerServiceImpl)runningService).schedule();

        verify(twitterClient, only()).startFetchProcess();
        verifyNoMoreInteractions(twitterClient);

    }
}

File: news-cast-explorer-plain/src/test/kotlin/org/jesperancinha/twitter/service/SchedulerServiceImplTest.kt
package org.jesperancinha.twitter.service

import com.ninjasquad.springmockk.MockkBean
import io.mockk.verify
import org.assertj.core.api.Assertions
import org.jesperancinha.twitter.client.TwitterClient
import org.junit.jupiter.api.Test
import org.mockito.Mockito
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.context.ActiveProfiles

@SpringBootTest
@ActiveProfiles("scheduler")
class SchedulerServiceImplTest(
    @Autowired
    val runningService: RunningService,
) {
    @MockkBean(relaxed = true)
    lateinit var twitterClient: TwitterClient

    @Test
    fun testStartProcess_whenRunnerCreated_thenSchedulerServiceInstance() {
        Assertions.assertThat(runningService).isInstanceOf(SchedulerServiceImpl::class.java)
    }

    @Test
    fun testStartProcess_whenRunnerStarted_thenSchedulerServiceStart() {
        runningService.startProcess()
        verify { twitterClient.startFetchProcess() }
    }

    @Test
    fun testStartProcess_whenSchedulerStarted_thenSchedulerServiceStarte() {
        (runningService as SchedulerServiceImpl?)!!.schedule()
        verify { twitterClient.startFetchProcess() }
    }
}